### DISCUSSION TITLE
Trigger on DSO symbols?

### DISCUSSION THREAD

**Discussion by mmirgttam** [janestreet/magic-trace#189] [Q&A] [answered]
I have an executable dynamically linked to several shared object libraries, and I can't seem to get magic-trace to find symbols in the linked libs, using either the fuzzy-finder with `-trigger ?` or just manually passing the mangled name to `-trigger <name>`. In the latter case, I get a `Snapshot symbol not found` error. The symbol I want to trigger on definitely exists in the DSO (I can find it with `nm -a <lib>`), and definitely does not appear in the `nm -a` output for the executable.

Is this expected, or is there maybe some issue with the way my libs are linked to my executable? 

I know this is a bit short on detail -- the code in question is proprietary -- but I wanted to ask first if I was just expecting something that wasn't possible.

**Comment by Xyene** [comment, marked as answer]
Hi! What you are asking for is not technically impossible, but it is not currently implemented in magic-trace.

There is no technical restriction preventing magic-trace from putting the breakpoint anywhere. For context, OCaml code (for which this tool was originally written) is typically all statically linked itself, so the only DSO functions of interest are functions in libc, etc. -- and you can trigger on the PLT trampolines for those.

The breakpoint is installed in https://github.com/janestreet/magic-trace/blob/a1807578fcb6a3e22f235125530e731f1a61cd8d/src/trace.ml#L248, if you were up for recompiling magic-trace from source you _could_ hardcode an address from a DSO there, or allow taking a raw hexadecimal address to `-trigger`. You'd lose the nicety of the fuzzy-finder, but it'd work.

As a more principled solution, one could imagine:

- for `magic-trace attach`, we could query what DSOs are mapped, and load up their symbols
- for `magic-trace run`, we could query what DSOs are linked, and load up their symbols
   - we couldn't handle `magic-trace run` with `dlopen`, because we wouldn't know beforehand what symbols are available

Patches would definitely be welcome :)
